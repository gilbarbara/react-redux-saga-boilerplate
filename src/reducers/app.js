import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';
import { REHYDRATE } from 'redux-persist/lib/constants';

import { ActionTypes } from 'constants/index';

export const appState = {
  alerts: [],
};

export default {
  app: handleActions(
    {
      [REHYDRATE]: state =>
        immutable(state, {
          alerts: { $set: [] },
        }),
      [ActionTypes.HIDE_ALERT]: (state, { payload: { id } }) => {
        const alerts = state.alerts.filter(d => d.id !== id);

        return immutable(state, {
          alerts: { $set: alerts },
        });
      },
      [ActionTypes.SHOW_ALERT]: (state, { payload }) =>
        immutable(state, {
          alerts: { $push: [payload] },
        }),
    },
    appState,
  ),
};
