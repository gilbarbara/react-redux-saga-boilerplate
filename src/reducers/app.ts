import { REHYDRATE } from 'redux-persist';

import { createReducer } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { AppState } from 'types';

export const appState: AppState = {
  alerts: [],
};

export default {
  app: createReducer<AppState>(
    {
      [REHYDRATE]: draft => {
        draft.alerts = [];
      },
      [ActionTypes.HIDE_ALERT]: (draft, { payload: { id } }) => {
        draft.alerts = draft.alerts.filter(d => d.id !== id);
      },
      [ActionTypes.SHOW_ALERT]: (draft, { payload }) => {
        draft.alerts.push(payload);
      },
    },
    appState,
  ),
};
