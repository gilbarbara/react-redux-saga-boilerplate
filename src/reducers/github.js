import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';
import { parseError } from 'modules/client';

import { ActionTypes, STATUS } from 'constants/index';

export const githubState = {
  repos: {
    data: {},
    status: STATUS.IDLE,
    message: '',
    query: '',
  },
};

export default {
  github: handleActions(
    {
      [ActionTypes.GITHUB_GET_REPOS]: (state, { payload }) => {
        const data = state.repos.data[payload.query] ? state.repos.data[payload.query] : [];

        return immutable(state, {
          repos: {
            data: {
              [payload.query]: { $set: data },
            },
            message: { $set: '' },
            query: { $set: payload.query },
            status: { $set: STATUS.RUNNING },
          },
        });
      },
      [ActionTypes.GITHUB_GET_REPOS_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          repos: {
            data: {
              [state.repos.query]: { $set: payload.data || [] },
            },
            status: { $set: STATUS.READY },
          },
        }),
      [ActionTypes.GITHUB_GET_REPOS_FAILURE]: (state, { payload }) =>
        immutable(state, {
          repos: {
            message: { $set: parseError(payload.message) },
            status: { $set: STATUS.ERROR },
          },
        }),
    },
    githubState,
  ),
};
