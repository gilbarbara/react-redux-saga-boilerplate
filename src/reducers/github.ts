import { parseError } from 'modules/client';
import { createReducer } from 'modules/helpers';

import { ActionTypes, STATUS } from 'literals';

import { GitHubState, Topic } from 'types';

const topic: Topic = {
  cached: false,
  data: [],
  message: '',
  status: STATUS.IDLE,
  updatedAt: 0,
};

export const githubState: GitHubState = {
  topics: {},
  query: '',
};

export default {
  github: createReducer<GitHubState>(
    {
      [ActionTypes.GITHUB_GET_REPOS_REQUEST]: (draft, { payload }) => {
        const { query } = payload;
        draft.query = query;

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].message = '';
        draft.topics[query].status = STATUS.RUNNING;
      },
      [ActionTypes.GITHUB_GET_REPOS_SUCCESS]: (draft, { meta, payload }) => {
        const { cached, query, updatedAt } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].cached = cached;
        draft.topics[query].data = payload;
        draft.topics[query].status = STATUS.SUCCESS;
        draft.topics[query].updatedAt = updatedAt;
      },
      [ActionTypes.GITHUB_GET_REPOS_FAILURE]: (draft, { meta, payload }) => {
        const { query } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].message = parseError(payload);
        draft.topics[query].status = STATUS.ERROR;
        draft.topics[query].updatedAt = 0;
      },
    },
    githubState,
  ),
};
