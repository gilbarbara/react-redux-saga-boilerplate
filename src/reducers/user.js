import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';

import { STATUS, ActionTypes } from 'constants/index';

export const userState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export default {
  user: handleActions(
    {
      [ActionTypes.USER_LOGIN]: state =>
        immutable(state, {
          status: { $set: STATUS.RUNNING },
        }),
      [ActionTypes.USER_LOGIN_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: true },
          status: { $set: STATUS.READY },
        }),
      [ActionTypes.USER_LOGOUT]: state =>
        immutable(state, {
          status: { $set: STATUS.RUNNING },
        }),
      [ActionTypes.USER_LOGOUT_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: false },
          status: { $set: STATUS.IDLE },
        }),
    },
    userState,
  ),
};
