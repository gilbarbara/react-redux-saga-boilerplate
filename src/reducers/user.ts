import { createReducer } from '@reduxjs/toolkit';

import { login, loginSuccess, logOut, logOutSuccess } from '~/actions';
import { STATUS } from '~/literals';

import { UserState } from '~/types';

export const userState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export default {
  user: createReducer<UserState>(userState, builder => {
    builder
      .addCase(login, draft => {
        draft.status = STATUS.RUNNING;
      })
      .addCase(loginSuccess, draft => {
        draft.isAuthenticated = true;
        draft.status = STATUS.READY;
      });

    builder
      .addCase(logOut, draft => {
        draft.status = STATUS.RUNNING;
      })
      .addCase(logOutSuccess, draft => {
        draft.isAuthenticated = false;
        draft.status = STATUS.IDLE;
      });
  }),
};
