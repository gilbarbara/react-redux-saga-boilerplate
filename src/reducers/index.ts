import app, { appState } from './app';
import github, { githubState } from './github';
import user, { userState } from './user';

export const initialState = {
  app: appState,
  github: githubState,
  user: userState,
};

export default {
  ...app,
  ...github,
  ...user,
};
