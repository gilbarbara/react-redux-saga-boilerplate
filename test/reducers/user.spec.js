import reducers from 'reducers';
import { ActionTypes } from 'constants/index';

describe('User', () => {
  it('should return the initial state', () => {
    expect(reducers.user(undefined, {})).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.USER_LOGIN_SUCCESS}`, () => {
    expect(reducers.user(undefined, { type: ActionTypes.USER_LOGIN_SUCCESS })).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.USER_LOGOUT_SUCCESS}`, () => {
    expect(reducers.user(undefined, { type: ActionTypes.USER_LOGOUT_SUCCESS })).toMatchSnapshot();
  });
});
