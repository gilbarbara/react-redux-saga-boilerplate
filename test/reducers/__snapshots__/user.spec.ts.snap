// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User > USER_LOGIN > should handle REQUEST 1`] = `
{
  "isAuthenticated": false,
  "status": "running",
}
`;

exports[`User > USER_LOGIN > should handle SUCCESS 1`] = `
{
  "isAuthenticated": true,
  "status": "ready",
}
`;

exports[`User > USER_LOGOUT > should handle REQUEST 1`] = `
{
  "isAuthenticated": false,
  "status": "running",
}
`;

exports[`User > USER_LOGOUT > should handle SUCCESS 1`] = `
{
  "isAuthenticated": false,
  "status": "idle",
}
`;

exports[`User > should return the initial state 1`] = `
{
  "isAuthenticated": false,
  "status": "idle",
}
`;
